contracts/PetPark.sol

```diff
@@ -1,7 +1,82 @@
 //SPDX-License-Identifier: Unlicense
 pragma solidity ^0.8.0;
 
-
 contract PetPark {
+    address owner;
+    enum AnimalTypes { None, Fish, Cat, Dog, Rabbit, Parrot}
+    enum Gender { Male, Female }
+
+    struct CallerInfo {
+        uint age;
+        Gender gender;
+        bool returningCaller;
+    }
+
+    mapping(uint => uint) public animalCounts;
+    mapping(address => uint) borrowedAnimalCount;
+    mapping(address => uint) borrowedAnimalType;
+    mapping(address => CallerInfo) callerNameAndAge;
+
+    event Added(uint indexed animalType, uint indexed count);
+    event Borrowed(uint indexed animalType);
+    event Returned(uint indexed animalType);
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    modifier onlyOwner() {
+        require(owner == msg.sender, "Not an owner");
+
+        _;
+    }
+
+    function add(uint _animalType, uint _count) external onlyOwner {
+        AnimalTypes currentType = AnimalTypes(_animalType);
+        require(AnimalTypes.None != currentType, "Invalid animal");
+
+        animalCounts[_animalType] += _count;
+
+        emit Added(_animalType, _count);
+    }
+
+    function borrow(uint _age, uint _gender, uint _animalType) external {
+        require(_age != 0, "Invalid Age");
+
+        if (callerNameAndAge[msg.sender].returningCaller) {
+            require(callerNameAndAge[msg.sender].age == _age, "Invalid Age");
```

Comment:
Since `callerNameAndAge[msg.sender].age` is being used multiple times it can be stored in a single variable. Later in course you will learn that this optimises gas usage